src/mem_table.rs

```diff
@@ -175,43 +175,49 @@ impl KVTable {
     /// Puts a value, returning as soon as the value is written to the memtable but before
     /// it is flushed to durable storage.
     fn put(&self, key: Bytes, value: Bytes, attrs: RowAttributes) {
-        self.maybe_subtract_old_val_from_size(key.clone());
         self.size.fetch_add(
             key.len() + value.len() + sizeof_attributes(&attrs),
             Ordering::Relaxed,
         );
-        self.map.insert(
+        let key_len = key.len();
+        self.map.compare_insert(
             key,
             ValueWithAttributes {
                 value: ValueDeletable::Value(value),
                 attrs,
             },
+            |old_value| {
+                self.subtract_old_val_from_size(key_len, old_value);
+                true
+            },
         );
     }
 
     fn delete(&self, key: Bytes, attrs: RowAttributes) {
-        self.maybe_subtract_old_val_from_size(key.clone());
         self.size
             .fetch_add(key.len() + sizeof_attributes(&attrs), Ordering::Relaxed);
-        self.map.insert(
+        let key_len = key.len();
+        self.map.compare_insert(
```

Comment:
often there's no guarantee that the function passed in to this type of concurrent data structure is not called multiple times (they typically don't have side effects), do you know if we have that guarantee here?

Reply:
I _think_ it's OK:

https://github.com/crossbeam-rs/crossbeam/blob/aba6f562d877be1c5b48df1b30afd3164a4c4464/crossbeam-skiplist/src/base.rs#L1083-L1099

Reply:
Maybe we can protect ourselves with a `OnceCell` anyway?

Reply:
I took a look inside and it wasn't totally clear to me that it won't be called more than once. It looks like the algorithm:

1. Passes in the function as `replace` into `insert_internal`
2. Attempts to find the first key, calls `replace` to see if it should be replaced. If it should, it continues
3. [Attempts to insert](https://github.com/crossbeam-rs/crossbeam/blob/aba6f562d877be1c5b48df1b30afd3164a4c4464/crossbeam-skiplist/src/base.rs#L933) it into the first layer of the skiplist. If it fails to do that, it searches for the key again and calls `replace` again. 

It's not clear to me why it might fail to insert, whether it's concurrency reasons or something else.

Reply:
Ah, it probably fails if there's a race condition where another the node at the found position was removed by another thread. Skiplist is lockfree. So, yea, it can definitely call replace multiple times. We need to guard against this. Really good catch.

Reply:
Great catch! I'll look into it 👍